### PR TITLE
User must logged in to view  premium content

### PR DESCRIPTION
**Pull request by designerofthing**
[PT Link](https://www.pivotaltracker.com/story/show/172888051)
[PT Link - chore](https://www.pivotaltracker.com/story/show/173008655)
```
As a User
In order to access the premium content
I would like to log in
```
- Link visitor able to log in
- Modify conditional rendering for user authentification
- Add login form as a component
- Sign up functionality for user
- Adding button log in/log out in the header
- Incl greeting user@mail and logout button to confirm user is logged in
- connect form to different requests
- Add JtockAuth
- setState authenticated using hooks (App component)

dnspr

dnsresponsive

### CODE REVIEW

File: src/components/Header.jsx [hunk 1]
```diff
@@ -1,15 +1,68 @@
...
-    <Grid columns={3} divided centered>
-      <Grid.Row>
-        <h1 id="header">Daily News Sense</h1>
-      </Grid.Row>
+const date = new Date();
+const currentTime = date.getHours();
+
+let time;
 
+if (currentTime < 12) {
```
Comment (kermit-klein) [inline src/components/Header.jsx:12]: LIKED THIS

File: src/components/LoginForm.jsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import React, { useState } from "react";
+import { Grid, Button, Form, Input } from "semantic-ui-react";
+import auth from "../modules/auth";
+import { useHistory } from "react-router-dom";
+import "../css/index.css";
+import { Link } from "react-router-dom";
+
+const LoginForm = (props) => {
+  const [message, setMessage] = useState("");
+  const history = useHistory();
```
Comment (kermit-klein) [inline src/components/LoginForm.jsx:10]: What does this do ?

File: cypress/integration/userCanLoginForPremiumContent.feature.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  it("with valid credentials", () => {
+    cy.get("div#login").within(() => {
+      cy.get("p").should("contain", "user@mail.com");
+    });
+  });
+
+  it("can view premium article", () => {
+    cy.visit("/article/3");
+    cy.get("#article-3-body").then((text) => {
+      text.length >= 100;
```
Comment (ViaMarcus) [inline cypress/integration/userCanLoginForPremiumContent.feature.js]: You should test for the presence of the PremiumBlocker component as well.
Both in happy path and sad path

File: cypress/fixtures/successful_login.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "success": true,
+  "data": {
+    "id": 1,
+    "email": "user@mail.com",
+    "provider": "email",
+    "uid": "user@mail.com",
+    "allow_password_change": false
+  }
```
Comment (ViaMarcus) [inline cypress/fixtures/successful_login.json:9]: the data will also include :role as soon as the API PR is merged, probably doesn't matter however and can be address when it does

File: src/components/Header.jsx [hunk 5]
```diff
@@ -1,15 +1,68 @@
...
+  time = "Evening";
+}
+
+const Header = (props) => {
+  const history = useHistory();
+  const logout = async () => {
+    try {
+      await auth.signOut();
+      props.setAuthenticated(false);
+      history.push("/");
```
Comment (ViaMarcus) [inline src/components/Header.jsx:26]: Is it really necessary to always redirect on logout?
Maybe only redirect if user is on a page they shouldn't be, such as their currently non-existing profile page.

File: src/components/Header.jsx [hunk 6]
```diff
@@ -1,15 +1,68 @@
...
+    }
+  };
+
+  return (
+    <Grid>
       <Grid.Row>
-        <Grid.Column></Grid.Column>
+        <Grid.Column>
+          <h1 id="header" style={{ textAlign: "center" }}>
+            <span>D</span>aily <span>N</span>ews <span>S</span>ense
```
Comment (ViaMarcus) [inline src/components/Header.jsx:37]: Neat, but very space consuming on mobile, according to screenshot.

File: src/components/LoginForm.jsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  const login = async (e) => {
+    try {
+      const response = await auth.signIn(
+        e.target.email.value,
+        e.target.password.value
+      );
+      if (response.success) {
+        props.setUid(response.data.uid);
+        props.setAuthenticated(true);
+        history.push("/");
```
Comment (ViaMarcus) [inline src/components/LoginForm.jsx]: Would it be possible to return the user to where they came from?
Maybe they visit a premium article, they realize they are not logged in, then they log in, and lose track of the article.

File: src/components/Navbar.jsx [hunk 8]
```diff
@@ -16,7 +16,7 @@ const Navbar = () => {
 
   let renderCategories = categories.map((cat) => {
     return (
-      <Menu.Item>
+      <Menu.Item boderless>
```
Comment (ViaMarcus) [inline src/components/Navbar.jsx:19]: borderless typo?

**Review by ViaMarcus — CHANGES_REQUESTED** [review]
A few tweaks may improve UX quite a bit

File: src/components/Header.jsx [hunk 9]
```diff
@@ -27,7 +40,7 @@ const Header = (props) => {
             </Link>
           ) : (
             <>
-              <p>Welcome {props.uid}</p>
+              <p>Good {time} {props.uid}</p>
```
Comment (PaulineBA) [inline src/components/Header.jsx]: Ouuuh That's fancy ;-) I like it!!

**Review by kermit-klein — APPROVED** [review]

File: cypress/integration/userCanLoginForPremiumContent.feature.js (hunk 3 above)
Comment (erikbjoern) [inline cypress/integration/userCanLoginForPremiumContent.feature.js]: The amount of characters in the snippet is updated to 300.

File: src/components/ArticleCard.jsx [hunk 10]
```diff
@@ -1,24 +1,26 @@
...
-          
-        > 
-          <h5 id="article-title" style={{ textAlign: "center" }}>{article.title}</h5>
+        >
+          <h5 id="article-title" style={{ textAlign: "center" }}>
+            {article.title}
+          </h5>
         </Link>
       </Placeholder>
     </Grid.Row>
```
Comment (erikbjoern) [inline src/components/ArticleCard.jsx]: Theses changes are out of scope, am I correct? The PR contains 18 files changed, so try to limit unnecessary changes.

File: src/components/ArticleList.jsx [hunk 11]
```diff
@@ -41,7 +41,7 @@ const ArticleList = (props) => {
 
   return (
     <div>
-      <Grid columns={3} divided centered>
+      <Grid fluid columns={3} divided centered>
         <Ad
           link={"https://www.mercedes-benz.com/en/"}
           id={"ad-1"}
```
Comment (erikbjoern) [inline src/components/ArticleList.jsx:50]: Same here, is this related to the feature in any way?

File: src/components/Header.jsx (hunk 1 above)
Comment (erikbjoern) [reply src/components/Header.jsx:12]: ME TOO!

File: src/components/Header.jsx [hunk 12]
```diff
@@ -1,15 +1,68 @@
...
+
+  return (
+    <Grid>
       <Grid.Row>
-        <Grid.Column></Grid.Column>
+        <Grid.Column>
+          <h1 id="header" style={{ textAlign: "center" }}>
+            <span>D</span>aily <span>N</span>ews <span>S</span>ense
+          </h1>
+          <div id="login">
```
Comment (erikbjoern) [inline src/components/Header.jsx:39]: I would like to see the login-button to be horizontally aligned with the title header. To reduce the height of the Header component.

File: src/css/Header.css [hunk 13]
```diff
@@ -1,23 +1,61 @@
...
+  color: white;
+  font-size: 40px;
+}
+
 h2,
-h3,
 h4,
 h5,
 h6 {
   font-family: "Quattrocento", serif;
```
Comment (erikbjoern) [inline src/css/Header.css:25]: Did you replace this font with Cinzel Decorative? Then font-family should be updated for the other headings as well.

**Review by erikbjoern — CHANGES_REQUESTED** [review]
Nice work guys! Found some things that could be changed however.

File: cypress/integration/userCanLoginForPremiumContent.feature.js [hunk 14]
```diff
@@ -0,0 +1,136 @@
+describe("successfully", () => {
```
Comment (oliverochman) [inline cypress/integration/userCanLoginForPremiumContent.feature.js]: Successfully what?

File: cypress/integration/userCanLoginForPremiumContent.feature.js [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+      cy.get("#password").type("password");
+      cy.get("Button#submit").contains("Submit").click();
+    });
+  });
+
+  it("and is directed to main page", () => {
+    cy.get("#article-1").should("contain", "title 1");
+  });
+
+  it("with valid credentials", () => {
```
Comment (oliverochman) [inline cypress/integration/userCanLoginForPremiumContent.feature.js]: what are you testing here?

File: cypress/integration/userCanLoginForPremiumContent.feature.js [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+  it("can view premium article", () => {
+    cy.visit("/article/3");
+    cy.get("#article-3-body").then((text) => {
+      text.length >= 100;
+    });
+  });
+});
+
+describe("unsuccessfully", () => {
```
Comment (oliverochman) [inline cypress/integration/userCanLoginForPremiumContent.feature.js]: everything in this file needs to be in the same describe block

File: src/components/ArticleCard.jsx (hunk 10 above)
Comment (jysmys) [reply src/components/ArticleCard.jsx]: The face lift of the styling.

File: src/components/ArticleList.jsx (hunk 11 above)
Comment (designerofthing) [reply src/components/ArticleList.jsx:50]: styling, trial and error

File: src/components/ArticleCard.jsx (hunk 10 above)
Comment (designerofthing) [reply src/components/ArticleCard.jsx]: it was part of the chore styling

File: cypress/integration/userCanLoginForPremiumContent.feature.js (hunk 3 above)
Comment (designerofthing) [reply cypress/integration/userCanLoginForPremiumContent.feature.js]: done, updated the fixture as well as it was 284 in total

File: src/components/Navbar.jsx (hunk 8 above)
Comment (designerofthing) [reply src/components/Navbar.jsx:19]: styling, chore

File: src/components/LoginForm.jsx (hunk 7 above)
Comment (designerofthing) [reply src/components/LoginForm.jsx]: we will look at that

File: cypress/integration/userCanLoginForPremiumContent.feature.js (hunk 3 above)
Comment (jysmys) [reply cypress/integration/userCanLoginForPremiumContent.feature.js]: that's in another user story, not this one i believe....

File: src/components/Header.jsx (hunk 6 above)
Comment (designerofthing) [reply src/components/Header.jsx:37]: it is the font size not the span, we will add to the responsiveness in Native I believe

File: cypress/fixtures/successful_login.json (hunk 4 above)
Comment (designerofthing) [reply cypress/fixtures/successful_login.json:9]: noted

File: src/components/Header.jsx (hunk 5 above)
Comment (designerofthing) [reply src/components/Header.jsx:26]: in general the log out is the last process of a user, they o not logout and then surf the same site some more with reduced access.  as convention we believe the redirect is good.

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: src/components/Navbar.jsx (hunk 8 above)
Comment (ViaMarcus) [reply src/components/Navbar.jsx:19]: But is it a typo and should say "borderless", because it doesn't

Comment (designerofthing) [reply src/components/Navbar.jsx:19]: yep, whoops, fixed!

File: cypress/integration/userCanLoginForPremiumContent.feature.js (hunk 14 above)
Comment (designerofthing) [reply cypress/integration/userCanLoginForPremiumContent.feature.js]: changed

File: cypress/integration/userCanLoginForPremiumContent.feature.js (hunk 15 above)
Comment (designerofthing) [reply cypress/integration/userCanLoginForPremiumContent.feature.js]: changed

File: src/components/ArticleCard.jsx (hunk 10 above)
Comment (erikbjoern) [reply src/components/ArticleCard.jsx]: I see, that's okay. Though we should probably avoid including chores in large PRs in the future.

File: cypress/integration/userCanLoginForPremiumContent.feature.js (hunk 3 above)
Comment (erikbjoern) [reply cypress/integration/userCanLoginForPremiumContent.feature.js]: Should we test that it doesn't appear when you're logged in?

File: src/components/Header.jsx (hunk 12 above)
Comment (designerofthing) [reply src/components/Header.jsx:39]: next story user can register we will address it

**Review by erikbjoern — APPROVED** [review]

File: src/components/PremiumBlocker.jsx [hunk 17]
```diff
@@ -1,18 +1,25 @@
-import React from 'react'
-import { Icon } from 'semantic-ui-react'
-import '../css/PremiumBlocker.css'
+import React from "react";
+import { Icon, Button } from "semantic-ui-react";
+import "../css/PremiumBlocker.css";
+import { Link, useHistory } from "react-router-dom";
```
Comment (oliverochman) [inline src/components/PremiumBlocker.jsx]: are you using `useHistory` for anything here?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: src/components/PremiumBlocker.jsx (hunk 17 above)
Comment (jysmys) [reply src/components/PremiumBlocker.jsx]: nope, it can be removed!
